File: app/src/main/java/com/example/kr_kotlin/ui/sub_catalog_java/ProductViewModel.java
package com.example.kr_kotlin.ui.sub_catalog_java;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ProductViewModel extends ViewModel {
    private final ProductRepository repository = new ProductRepository();

    private final MutableLiveData<List<Product>> _products = new MutableLiveData<>();
    public LiveData<List<Product>> getProducts() {
        return _products;
    }

    private final MutableLiveData<List<Product>> _favorites = new MutableLiveData<>();
    public LiveData<List<Product>> getFavorites() {
        return _favorites;
    }

    public void fetchProducts(String category, String search) {
        repository.getProducts(category, search, productList -> {
            if (search != null && !search.isEmpty()) {
                List<Product> filtered = productList.stream()
                        .filter(p -> p.getName().toLowerCase().contains(search.toLowerCase()))
                        .collect(Collectors.toList());
                _products.setValue(filtered);
            } else {
                _products.setValue(productList);
            }
        });
    }

    public void toggleFavorite(Product product) {
        List<Product> currentFavs = _favorites.getValue();
        boolean isFav = currentFavs != null &&
                currentFavs.stream().anyMatch(p -> p.getId().equals(product.getId()));

        if (isFav) {
            repository.removeFromFavorites(product.getId(), this::loadFavorites);
        } else {
            repository.addToFavorites(product, this::loadFavorites);
        }
    }

    public void loadFavorites() {
        repository.getFavorites(products -> _favorites.setValue(products));
    }
}

File: app/src/main/java/com/example/kr_kotlin/ui/sub_catalog/ProductViewModel.kt
package com.example.kr_kotlin.ui.sub_catalog

import android.os.Parcelable
import android.widget.ImageView
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.kr_kotlin.R
import com.example.kr_kotlin.ui.cart.CartViewModel
import kotlinx.parcelize.Parcelize

@Parcelize
class ProductViewModel : ViewModel(), Parcelable {
    private val repository = ProductRepository()


    private val _products = MutableLiveData<List<Product>>()
    val products: LiveData<List<Product>> get() = _products

    private val _favorites = MutableLiveData<List<Product>>()
    val favorites: LiveData<List<Product>> get() = _favorites

    fun fetchProducts(category: String? = null, search: String? = null) {
        repository.getProducts(category, search) { productList ->
            if (search != null) {
                _products.value = productList.filter {
                    it.name.contains(search, ignoreCase = true)
                }
            } else {
                _products.value = productList
            }
        }
    }


    fun toggleFavorite(product: Product) {
        val isFav = _favorites.value?.any {it.id == product.id} == true

        if (isFav) {
            repository.removeFromFarovires(product.id) {
                loadFavorites()
            }
        }
        else {
            repository.addToFavorites(product) {
                loadFavorites()
            }
        }

    }

    fun loadFavorites() {
        repository.getFavotites { _favorites.value = it }
    }
}